File: Java/938_Range_Sum_Of_BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int rangeSumBST(TreeNode root, int low, int high) {
        //base case
        if(root == null) {
            return 0;
        }
        //if both values are in range we add the value and check both
        //subtrees
        if(root.val <= high && root.val >= low) {
            return root.val + rangeSumBST(root.left,low,high)
            + rangeSumBST(root.right, low, high);
            //otherwise if root is less or equal to high we check right
        } else if(root.val <= high) {
            return rangeSumBST(root.right, low,high);
        } else {
            //otherwise we check left subtree as root is more than low
            return rangeSumBST(root.left,low,high);
        } 
    }
}

File: C/938_Range_Sum_Of_BST.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int rangeSumBST(struct TreeNode* root, int low, int high) {
    
    int toAdd = 0;
    //check to see if root is null as our base case
    if(!root) {
        return 0;
        //if the roots value is in our range we add it to our solution
    } else if(root->val <= high && root->val >= low) {
        toAdd = root->val;
    } 
    //add the value we need and recursively check both subtrees
    return toAdd + rangeSumBST(root->left,low,high) + rangeSumBST(root->right, low, high);

}
